### ISSUE TITLE
Passing RenderSection as the model parameter of RenderPartial does not render in the correct place in MVC6

### ISSUE THREAD

**Issue by JosKrause** [aspnet/Mvc#3813]
In MVC5 I was able to use the following construct:

``` csharp
@{ Html.RenderPartial("_Header", RenderSection("topright", false)); }
```

This allowed me to override the section inside the partial:

``` csharp
<header>
        <!-- insert jaw-dropping markup here for the left container -->

        @if (@Model is HelperResult)
        {
            @Model  <!-- This would render the contents of the @section -->
        }
        else
        {
            <div class="pull-right">
                @Html.Partial("_TopRight")
            </div>
        }
</header>
```

Now whether this approach is actually how it should be done, I will not debate the outcome, just because this worked in MVC5 I will not expect this to magically work in MVC6 (I know `RenderPartial` is being re-added in RC2, that's great) but I am trying to understand why this construct is no longer working in MVC6 and if this approach is deprecated or very evil, wondering how this could be achieved using MVC6 ;-)

When I check the debugger I can see that a dynamic (`HtmlString`) is being passed as the value of `@Model` but the contents are empty. So obviously something is different now with MVC6, I can't make it click so any guidance on how to solve this particular issue would be appreciated (debugger sometimes also indicates that the section is already rendered, weird).

Some extra background in why I have used this approach in MVC5:
- I have a `<header>` element that contains two div containers, one for the left and one for the right
- I created a `PartialView` that contains all the header mark-up
- I cannot use `RenderSection` in _Layout since this would render it outside the </header> tag thus borking my style (right?)
- By explicitly checking for `HelperResult` I could make sure that only if the `@section` was actually defined on the View and not some other `Model` instance was being passed it would render the contents of that section, if this was not the case then it would output the default content and no harm done
- **This would thus enable me to override the section content when defined and display default content when the section is not defined (so no override)**

What struck me as odd is that `RenderSection` does seem to be called and the contents are rendered but they are now placed **before** the contents of the partial view (in my case the content of the section is injected before the `<header>` tag).

**TL;DR:** Is it intended that this approach is no longer working in MVC6? If so, what would be the best approach in achieving a similar result. If not, would this be added back when `RenderPartial` is resstored in RC2?

**Comment by pranavkm** [comment]
The behavior of `RenderSection` is a bit strange primarily to support rendering sections asynchronously (`RenderSectionAsync`). When you invoke `RenderSection`, we write to `RazorPage.Output` asynchronously and return a token value that doesn't represent the content that is rendered but lets the   `Write` call that comes with `@RenderSection("foo")` to behave. The rationale for this approach was to allow async Flush calls made inside a section to work asynchronously rather than block.

As of this point, the only way to get section content in a variable would be to replace the `Output`. 
```C#
private IHtmlContent RenderSectionToString()
{
    var oldWriter = ViewContext.Output;
    using (var writer = new StringWriter())
    {  
        ViewContext.Output = writer;
        RenderSection("topRight", false);
        ViewContext.Output = oldWriter;
        
        return new HtmlString(writer.ToString());
    }
}
```

You might be able to reuse some of the TagHelper infrastructure for this:
    
public IHtmlContent RenderSectionToString()
{
    StartTagHelperWritingScope();
    RenderSection("topRight", false);
    return EndTagHelperWritingScope();
}
```

Our design choices might require revisiting since the behavior isn't particularly easy to follow.

**Comment by Eilon** [comment]
Moving to backlog to remember to reconsider this in the future, in case we see this as a larger problem.

For now I'd use the workaround that @pranavkm mentioned, though it's hacky and probably not quite "supported".

To be honest, I'm a bit surprised this *ever* worked in MVC/Razor because it certainly wasn't an intentional behavior.

**Comment by JosKrause** [comment]
Thanks so much for the provided feedback and valuable insights as to how this all fits together. For the time being I have resorted into using the following approach:

'''
    <header id="header">
        @{ await Html.RenderPartialAsync("_Header"); }
        <section id="top-right">
            @(IsSectionDefined("topright") ? RenderSection("topright", false) : Html.Partial("_TopRight"))
        </section>
    </header>
'''

For now it is giving me the result I need, but being able to define default content when a section is not defined is not an uncommon scenario; will see in the upcoming builds what remains possible to achieve this, looking forward!

**Comment by mkArtakMSFT** [comment]
Thanks for contacting us. We believe that the question you've raised have been answered. If you still feel a need to continue the discussion, feel free to reopen it and add your comments.
